fvtest/compilertriltest/TypeConversionTest.cpp

```diff
@@ -908,3 +908,386 @@ INSTANTIATE_TEST_CASE_P(TypeConversionTest, Int64ToDouble, ::testing::Combine(
     ::testing::Values(
         std::make_tuple("l2d", l2d) )));
 #endif
+
+template <typename T>
+bool smallFp_filter(T a)
+   {
+   // avoid failure caused by snprintf("%f")
+   return (abs(a) < 0.01);
+   }
+
+int32_t f2i(float x) {
+    return static_cast<int32_t>(x);
+}
+
+int64_t f2l(float x) {
+    return static_cast<int64_t>(x);
+}
+
+int32_t d2i(double x) {
+    return static_cast<int32_t>(x);
+}
+
+int64_t d2l(double x) {
+    return static_cast<int64_t>(x);
+}
+
+class FloatToInt32 : public TRTest::UnaryOpTest<int32_t,float> {};
+
+TEST_P(FloatToInt32, UsingConst) {
+    auto param = TRTest::to_struct(GetParam());
+
+    char inputTrees[160] = {0};
+    std::snprintf(inputTrees, 160,
+        "(method return=Int32"
+        "  (block"
+        "    (ireturn"
+        "      (f2i"
+        "        (fconst %f) ) ) ) )",
+        param.value);
+    auto trees = parseString(inputTrees);
+
+    ASSERT_NOTNULL(trees);
+
+    Tril::DefaultCompiler compiler{trees};
+
+    ASSERT_EQ(0, compiler.compile()) << "Compilation failed unexpectedly\n" << "Input trees: " << inputTrees;
+
+    auto entry_point = compiler.getEntryPoint<int32_t (*)()>();
+    volatile auto exp = param.oracle(param.value);
+    volatile auto act = entry_point();
+    ASSERT_EQ(exp, act);
+}
+
+TEST_P(FloatToInt32, UsingLoadParam) {
+    auto param = TRTest::to_struct(GetParam());
+
+    char *inputTrees =
+        "(method return=Int32 args=[Float]"
+        "  (block"
+        "    (ireturn"
+        "      (f2i"
+        "        (fload parm=0) ) ) ) )";
+    auto trees = parseString(inputTrees);
+
+    ASSERT_NOTNULL(trees);
+
+    Tril::DefaultCompiler compiler{trees};
+
+    ASSERT_EQ(0, compiler.compile()) << "Compilation failed unexpectedly\n" << "Input trees: " << inputTrees;
+
+    auto entry_point = compiler.getEntryPoint<int32_t (*)(float)>();
+    volatile auto exp = param.oracle(param.value);
+    volatile auto act = entry_point(param.value);
+    ASSERT_EQ(exp, act);
+}
+
+INSTANTIATE_TEST_CASE_P(TypeConversionTest, FloatToInt32, ::testing::Combine(
+    ::testing::ValuesIn(
+        TRTest::filter(TRTest::const_values<float>(), smallFp_filter<float>)),
```

Comment:
Although I agree it makes sense to use a template function for `smallFp_filter`, I believe some of our older build compiler's don't handle converting template function instantiations to function pointers correctly. So, a separate function needs to be defined for each type instead :disappointed: . (This is also the reason template functions are not used for the oracle functions)

Reply:
Of course I can change `smallFp_filter()` not to use template.
But fvtest/compilertriltest/ArithmeticTest.cpp already uses a filter function with template:
```
        TRTest::filter(TRTest::const_value_pairs<int32_t, int32_t>(), div_filter<int32_t> )),
```
Does the `div_filter()` cause problems anywhere?

Reply:
Hm... so it does. Ok, I'll accept the changes as they are then.

I'll need to take a closer look at what was not working before. If this does turn out to be a problem, I will fix it when there's concrete evidence.